lib/pos_fee_estimator.go

```diff
@@ -194,19 +210,44 @@ func (posFeeEstimator *PoSFeeEstimator) removeBlockNoLock(block *MsgDeSoBlock) e
 func (posFeeEstimator *PoSFeeEstimator) UpdateGlobalParams(globalParams *GlobalParamsEntry) error {
 	posFeeEstimator.rwLock.Lock()
 	defer posFeeEstimator.rwLock.Unlock()
+
+	// Create a temporary transaction register to test the new global params.
 	tempTransactionRegister := NewTransactionRegister()
 	tempTransactionRegister.Init(globalParams.Copy())
+
 	for _, block := range posFeeEstimator.cachedBlocks {
 		if err := addBlockToTransactionRegister(tempTransactionRegister, block); err != nil {
 			return errors.Wrap(err, "PosFeeEstimator.UpdateGlobalParams: error adding block to tempTransactionRegister")
 		}
 	}
+
+	posFeeEstimator.globalParams = globalParams
 	return nil
 }
 
-// sortCachedBlocks sorts the cached blocks by height & tstamp just to be safe.
-func (posFeeEstimator *PoSFeeEstimator) sortCachedBlocks() {
-	posFeeEstimator.cachedBlocks = collections.SortStable(posFeeEstimator.cachedBlocks,
+// cleanUpPastBlocks cleans up the input blocks slice, deduping, sorting, and pruning the blocks by height.
+func (posFeeEstimator *PoSFeeEstimator) cleanUpPastBlocks(blocks []*MsgDeSoBlock) []*MsgDeSoBlock {
+	dedupedBlocks := posFeeEstimator.dedupeBlocksByBlockHeight(blocks)
+	sortedBlocks := posFeeEstimator.sortBlocksByBlockHeightAndTstamp(dedupedBlocks)
+	return posFeeEstimator.pruneBlocksToMaxNumPastBlocks(sortedBlocks)
+}
+
+// dedupeBlocksByBlockHeight deduplicates the blocks by block height. If multiple blocks have the same
+// height, it keeps the one with the highest view.
+func (posFeeEstimator *PoSFeeEstimator) dedupeBlocksByBlockHeight(blocks []*MsgDeSoBlock) []*MsgDeSoBlock {
+	blocksByBlockHeight := make(map[uint64]*MsgDeSoBlock)
+	for _, block := range blocks {
+		existingBlock, hasExitingBlock := blocksByBlockHeight[block.Header.Height]
+		if !hasExitingBlock || existingBlock.Header.GetView() < block.Header.GetView() {
+			blocksByBlockHeight[block.Header.Height] = block
+		}
+	}
+	return collections.MapValues(blocksByBlockHeight)
+}
+
+// sortBlocksByBlockHeightAndTstamp sorts the blocks by height & tstamp just to be safe.
+func (posFeeEstimator *PoSFeeEstimator) sortBlocksByBlockHeightAndTstamp(blocks []*MsgDeSoBlock) []*MsgDeSoBlock {
+	return collections.SortStable(blocks,
 		func(ii, jj *MsgDeSoBlock) bool {
 			if ii.Header.Height != jj.Header.Height {
```

Comment:
With these updates I think we should no longer have blocks at the same height?

Reply:
Yes, you're right. We don't need the tie-breaking anymore. Maybe it's safer to keep it regardless? This is a pretty robust sorting function that will work if the the deduping functionality elsewhere in this file gets changed